### PR TITLE
TST: Add test decorators for redirecting stdout and stderr

### PR DESCRIPTION
**Pull request by gfyoung**
Add testing decorators for redirecting `stdout` and `stderr`.

xref <a href="https://github.com/pandas-dev/pandas/pull/15925#discussion_r110283696">#15925 (comment)</a>

### CODE REVIEW

File: pandas/tests/io/parser/common.py [hunk 1]
```diff
@@ -1292,16 +1290,16 @@ def test_verbose_import(self):
...
-                self.assertEqual(buf.getvalue(),
-                                 'Filled 1 NA values in column a\n')
-        finally:
-            sys.stdout = sys.__stdout__
+        self.read_csv(StringIO(text), verbose=True, index_col=0)
+        output = sys.stdout.getvalue()
+
+        # Engines are verbose in different ways.
+        if self.engine == 'c':
+            self.assertIn('Tokenization took:', output)
```
Comment (jreback) [inline pandas/tests/io/parser/common.py]: while you are at it, can you use pytest idioms here, e.g. ``self.assertIn`` -> ``assert output in ......``

File: pandas/tests/io/parser/common.py [hunk 2]
```diff
@@ -1292,16 +1290,16 @@ def test_verbose_import(self):
...
+            self.assertIn('Parser memory cleanup took:', output)
+        else:  # Python engine
+            exp = 'Filled 1 NA values in column a\n'
+            self.assertEqual(output, exp)
```
Comment (jreback) [inline pandas/tests/io/parser/common.py]: assert output == exp

File: pandas/tests/io/test_sql.py [hunk 3]
```diff
@@ -2259,12 +2256,10 @@ def test_execute_closed_connection(self):
 
         sql.execute('INSERT INTO test VALUES("foo", "bar", 1.234)', self.conn)
         self.conn.close()
-        try:
-            sys.stdout = StringIO()
-            self.assertRaises(Exception, tquery, "select * from test",
-                              con=self.conn)
-        finally:
-            sys.stdout = sys.__stdout__
+
```
Comment (jreback) [inline pandas/tests/io/test_sql.py]: use pytest.raises

File: pandas/util/testing.py [hunk 4]
```diff
@@ -2790,3 +2790,37 @@ def setTZ(tz):
         yield
     finally:
         setTZ(orig_tz)
+
+
+def capture_stdout(f):
+    """
+    Capture stdout in a buffer so that it can be checked
+    (or suppressed) during testing.
+    """
```
Comment (jreback) [inline pandas/util/testing.py]: can you say this is a decorator, and show how to use it. (same as on capture_stderr).

Comment (jreback) [inline pandas/util/testing.py]: also move these near the *other* routine which maniuplates stdout (about encoding IIRC).

**Review by jreback — APPROVED** [review]

File: pandas/tests/io/parser/common.py (hunk 1 above)
Comment (gfyoung) [inline pandas/tests/io/parser/common.py]: Sure thing.  Done.

File: pandas/tests/io/parser/common.py (hunk 2 above)
Comment (gfyoung) [inline pandas/tests/io/parser/common.py]: Yep, done.

File: pandas/tests/io/test_sql.py (hunk 3 above)
Comment (gfyoung) [inline pandas/tests/io/test_sql.py]: Yep, done.

File: pandas/util/testing.py (hunk 4 above)
Comment (gfyoung) [inline pandas/util/testing.py]: Yep, done.

**Comment by jreback** [comment]
thanks @gfyoung !

**Pull request merged**